### PR TITLE
Filter for Date range and Location range from search bar

### PR DESCRIPTION
**Pull request by JakeThurman**
**Note:** The following is not included, this is a separate story. (The filter sliders will persist only until you leave the view)

> The "distance" and "date" filters in the search component should be persistent across all instances of the search component

To Test:
1. Navigate to Topics Tab
2. Select any Topic
3. Tap the "Filter" button on search bar
4. Drag sliders, and observe changes in search results after tapping "Filter"

### CODE REVIEW

File: Buddies/Activities/ActivityTableVC.swift [hunk 1]
```diff
@@ -38,6 +33,11 @@ class ActivityTableVC: UITableViewController, FilterSearchBarDelegate {
         fab.renderCreateActivityFab()
```
Comment (noahtallen) [inline Buddies/Activities/ActivityTableVC.swift:33]: ;)

File: Buddies/Controls/FilterSearchBar.swift [hunk 2]
```diff
@@ -18,16 +19,21 @@ protocol FilterSearchBarDelegate {
...
+
     var api: AlgoliaSearch //Should only be changed by unit tests
-    
     var displayDelegate: FilterSearchBarDelegate?
     
-    // Default is a sentinal because XCode hates nil tuples :(
-    var lastSearchParams: SearchParams = ("", nil, 0)
+    // This just uses default values for now
+    //  these probably need to be stored in
+    //  firestore for BUD-41 🤔
```
Comment (noahtallen) [inline Buddies/Controls/FilterSearchBar.swift:29]: Probably easier to do that than using local storage lol

File: Buddies/Controls/FilterSearchBar.swift [hunk 3]
```diff
@@ -47,87 +53,200 @@ class FilterSearchBar : UISearchBar, UISearchBarDelegate {
...
-        renderFilterButton()
+        let bttn = makeFilterButton(
+            doing: #selector(self.onFilterTapped),
+            rounding: [.layerMaxXMaxYCorner, .layerMaxXMinYCorner])
+        
+        addSubview(bttn)
+        
+        // Size/Position
+        NSLayoutConstraint.activate([
+            bttn.rightAnchor.constraint(equalTo: rightAnchor, constant: -10),
```
Comment (noahtallen) [inline Buddies/Controls/FilterSearchBar.swift:66]: Any reason why these are negative?

File: Buddies/Controls/FilterSearchBar.swift [hunk 4]
```diff
@@ -47,87 +53,200 @@ class FilterSearchBar : UISearchBar, UISearchBarDelegate {
...
+            dateSlider,
+            locationSliderLabel,
+            locationRangeSlider,
+            innerFilterButton,
+            cancelButton,
+        ]
+        container.addSubview(blurView)
+        containerChildren.forEach { container.addSubview($0) }
+        
+        let parent = superview!.superview! // Get the window in an ugly way
```
Comment (noahtallen) [inline Buddies/Controls/FilterSearchBar.swift]: I also didn't like having to do this - It's like "fair" for like one level

Comment (noahtallen) [reply Buddies/Controls/FilterSearchBar.swift]: Pretty much impossible to avoid at this point, though

Comment (noahtallen) [reply Buddies/Controls/FilterSearchBar.swift]: oh this was supposed to go in my review lol. oops.

File: Buddies/Controls/DateRangeSliderDelegate.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
+//
```
Comment (noahtallen) [inline Buddies/Controls/DateRangeSliderDelegate.swift:1]: Think this could be integrated with create activity down the road?

File: Buddies/Controls/FilterSearchBar.swift (hunk 4 above)
Comment (noahtallen) [reply Buddies/Controls/FilterSearchBar.swift]: oh and apparently all my other comments went singularly as well when I posted this one. oh well

File: Buddies/Controls/DateRangeSliderDelegate.swift (hunk 5 above)
Comment (noahtallen) [reply Buddies/Controls/DateRangeSliderDelegate.swift:1]: I see you've already done this ;)

File: Buddies/Controls/FilterSearchBar.swift [hunk 6]
```diff
@@ -152,12 +275,16 @@ class FilterSearchBar : UISearchBar, UISearchBarDelegate {
         // Store request params #NoRaceConditions
         self.lastSearchParams = myParams
         
+        let start = DateRangeSliderDelegate.instance.getSliderDate(sliderValue: CGFloat(myParams.dateMin))
```
Comment (noahtallen) [inline Buddies/Controls/FilterSearchBar.swift]: Seems a bit wordy - would there be a way to clean this up? (Not quite sure how the architecture is working here)

File: Buddies/Controls/FilterSearchBar.swift [hunk 7]
```diff
@@ -152,12 +275,16 @@ class FilterSearchBar : UISearchBar, UISearchBarDelegate {
...
+        
         // Load data from algolia!
         api.searchActivities(withText: myParams.filterText,
                              matchingAnyTopicOf: displayDelegate?.getTopics(),
-                             startingAt: myParams.when?.start,
-                             endingAt: myParams.when?.end,
-                             upToDisatnce: myParams.maxMetersAway) {
+                             startingAt: start,
+                             endingAt: end,
+                             upToDisatnce: distance) {
```
Comment (noahtallen) [inline Buddies/Controls/FilterSearchBar.swift:277]: awesome! at some point we'll have to assign someone to get the actual user's current location... it ignores `upToDistance` if you don't specify coordinates. Not an issue for this PR though

**Review by noahtallen — CHANGES_REQUESTED** [review]
It looks good! everything seems to work well, but there is a small bug that we should probably fix before merging. Basically, you can swipe on the left edge of the screen to navigate back. However, the buttons of the sliders fall into that area, so when they're at their lowest values, trying to move the sliders closes the filter menu and navigates back instead. See gif for clarification, lmk if you have questions about it!

swipe_bug

**Comment by noahtallen** [comment]
screen shot 2019-02-24 at 6 34 30 pm

It would also be nice if the margin at the top was a bit bigger. Maybe ~20?

**Comment by JakeThurman** [comment]
TODO: fix build error from tests (oops)

```
❌  /Users/travis/build/jack-and-the-beans/buddies-ios/BuddiesTests/Controls/FilterSearchBarTests.swift:71:9: value of type 'FilterSearchBar?' has no member 'searchBarCancelButtonClicked'
        bar.searchBarCancelButtonClicked(bar)
                ^
```

**Comment by JakeThurman** [comment]
The margin wasn't like then when I developed on an iPhone 7. Obviously, that looks bad. I'll look into it!

File: Buddies/Controls/FilterSearchBar.swift (hunk 3 above)
Comment (JakeThurman) [reply Buddies/Controls/FilterSearchBar.swift:66]: I don't really know but they overlapped when it was positive and they looked correct when it was this direction.

Comment (noahtallen) [reply Buddies/Controls/FilterSearchBar.swift:66]: cool

File: Buddies/Controls/FilterSearchBar.swift (hunk 6 above)
Comment (JakeThurman) [reply Buddies/Controls/FilterSearchBar.swift]: Definetly. I wasn't sure how this method was being used and didn't fix it. I think I'm going to change it to be called like this:

```swift
DateRangeSliderDelegate.getDate(sliderIndex: myParams.dateMin)
```

I think that looks much better. Thanks for calling that out!

Comment (noahtallen) [reply Buddies/Controls/FilterSearchBar.swift]: cool!

**Comment by noahtallen** [comment]
FYI one helpful constraint function I added as an extension to UIView binds the UIView frame to the edges of the superview. Not sure if that would be helpful for you

**Comment by JakeThurman** [comment]
I just rebased, and fixed remaining issues 👍

**Review by noahtallen — APPROVED** [review]
